### PR TITLE
Handle colons in file names when producing SemanticDB

### LINKED ISSUE #15860: SemanticDB URISyntaxException
I have a source file whose name contains a colon, where the part before the colon is not a valid URI scheme according to RFC 2396. When I compile with SemanticDB enabled, compilation fails with a `URISyntaxException`.

## Compiler version

3.2.0-RC3, 3.1.3

## Minimized code

In a shell:
```sh
touch 1:a.scala
cs launch scalac:3.2.0-RC3 -- -Ysemanticdb 1:a.scala
```

## Output

```
exception occurred while compiling 1:a.scala
java.net.URISyntaxException: Illegal character in scheme name at index 0: 1:a.scala while compiling 1:a.scala
Exception in thread "main" java.net.URISyntaxException: Illegal character in scheme name at index 0: 1:a.scala
	at java.base/java.net.URI$Parser.fail(URI.java:2973)
	at java.base/java.net.URI$Parser.checkChars(URI.java:3144)
	at java.base/java.net.URI$Parser.checkChar(URI.java:3154)
	at java.base/java.net.URI$Parser.parse(URI.java:3169)
	at java.base/java.net.URI.<init>(URI.java:708)
	at java.base/java.net.URI.<init>(URI.java:809)
	at dotty.tools.dotc.semanticdb.Tools$.$anonfun$1(Tools.scala:15)
	at scala.collection.StrictOptimizedIterableOps.map(StrictOptimizedIterableOps.scala:100)
	at scala.collection.StrictOptimizedIterableOps.map$(StrictOptimizedIterableOps.scala:87)
	at scala.collection.convert.JavaCollectionWrappers$JIterableWrapper.map(JavaCollectionWrappers.scala:58)
	at dotty.tools.dotc.semanticdb.Tools$.mkURIstring(Tools.scala:15)
	at dotty.tools.dotc.semanticdb.ExtractSemanticDB$.write(ExtractSemanticDB.scala:499)
	at dotty.tools.dotc.semanticdb.ExtractSemanticDB.run(ExtractSemanticDB.scala:50)
	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:311)
	at scala.collection.immutable.List.map(List.scala:246)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:312)
	at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:234)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1328)
	at dotty.tools.dotc.Run.runPhases$1(Run.scala:245)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:253)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:262)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:68)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:262)
	at dotty.tools.dotc.Run.compileSources(Run.scala:186)
	at dotty.tools.dotc.Run.compile(Run.scala:170)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:35)
	at dotty.tools.dotc.Driver.process(Driver.scala:195)
	at dotty.tools.dotc.Driver.process(Driver.scala:163)
	at dotty.tools.dotc.Driver.process(Driver.scala:175)
	at dotty.tools.dotc.Driver.main(Driver.scala:205)
	at dotty.tools.MainGenericCompiler$.run$1(MainGenericCompiler.scala:162)
	at dotty.tools.MainGenericCompiler$.main(MainGenericCompiler.scala:186)
	at dotty.tools.MainGenericCompiler.main(MainGenericCompiler.scala)
```

### PR DESCRIPTION
**Pull request by kierendavies**
Fixes #15860

This solution feels pretty hacky to me, but it passes all existing tests... :shrug:

### CODE REVIEW

**Comment by kierendavies** [comment]
Apparently in CI it does _not_ pass all existing tests. I'm not sure how to interpret the job failure messages though. Can someone help me out?

**Comment by bishabosha** [comment]
> Apparently in CI it does _not_ pass all existing tests. I'm not sure how to interpret the job failure messages though. Can someone help me out?

something changed in the output of semanticdb - you can test if this was intentional by running
```
sbt 'scala3-compiler-bootstrapped/test:runMain dotty.tools.dotc.semanticdb.updateExpect'
```
which will update some "expect" files, (used regression testing semanticdb output) and then comparing the diff with git.

**Comment by kierendavies** [comment]
Thank you @bishabosha. It was a bug, which I've fixed.

File: compiler/src/dotty/tools/dotc/semanticdb/Tools.scala [hunk 1]
```diff
@@ -12,7 +12,11 @@ object Tools:
   def mkURIstring(path: Path): String =
     // Calling `.toUri` on a relative path will convert it to absolute. Iteration through its parts instead preserves
     // the resulting URI as relative.
-    val uriParts = for part <- path.asScala yield new java.net.URI(null, null, part.toString, null)
+    // To prevent colon `:` from being treated as a scheme separator, prepend a slash `/` to trick the URI parser into
+    // treating the part as an absolute path, and then remove it afterwards.
+    val uriParts =
+      for part <- path.asScala
+      yield new java.net.URI(null, null, "/" + part.toString, null).toString().stripPrefix("/")
     uriParts.mkString("/")
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/semanticdb/Tools.scala:20]: would it have the same result if you don't strip each individual part, and instead join the parts with `""` and then only strip the first prefix

**Review by bishabosha — APPROVED** [review]

**Comment by bishabosha** [comment]
hopefully CI passes, thank you!

File: compiler/src/dotty/tools/dotc/semanticdb/Tools.scala [hunk 2]
```diff
@@ -12,8 +12,10 @@ object Tools:
   def mkURIstring(path: Path): String =
     // Calling `.toUri` on a relative path will convert it to absolute. Iteration through its parts instead preserves
     // the resulting URI as relative.
-    val uriParts = for part <- path.asScala yield new java.net.URI(null, null, part.toString, null)
-    uriParts.mkString("/")
+    // To prevent colon `:` from being treated as a scheme separator, prepend a slash `/` to each part to trick the URI
+    // parser into treating it as an absolute path, and then strip the spurious leading slash from the final result.
+    val uriParts = for part <- path.asScala yield new java.net.URI(null, null, "/" + part.toString, null)
```
Comment (ritschwumm) [inline compiler/src/dotty/tools/dotc/semanticdb/Tools.scala:17]: looks like this URI constructor is just not the right tool  for the job...

Comment (kierendavies) [reply compiler/src/dotty/tools/dotc/semanticdb/Tools.scala:17]: It's a poor fit for sure, but sadly the Java standard library doesn't provide anything better. I considered copy-pasting the OpenJDK private method on `java.net.URI` that does the quoting, but it's, uh, [_very_ Java](https://github.com/openjdk/jdk/blob/802ef38060080254e55621e4c64fa31a6c0b7b18/src/java.base/share/classes/java/net/URI.java#L2772).

Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/semanticdb/Tools.scala:17]: its the only way to parse URI as the URI parser is private to the class

**Pull request merged**